### PR TITLE
[Enhancement] Support materialized_view_rewrite_mode for mv rewrite

### PR DESCRIPTION
**Pull request by LiShuMing**
Fixes #issue

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 111 / 131 (84.73%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|0|3|00.00%|[137, 138, 139]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|7|10|70.00%|[188, 381, 518]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|18|25|72.00%|[162, 163, 167, 220, 263, 266, 267]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/OptimizeGroupTask.java|5|6|83.33%|[46]|
|:large_blue_circle:|com/starrocks/sql/optimizer/GECost.java|11|13|84.62%|[36, 45]|
|:large_blue_circle:|com/starrocks/sql/optimizer/GroupExpression.java|6|7|85.71%|[292]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|14|16|87.50%|[891, 905]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|9|10|90.00%|[1792]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/SortProperty.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Memo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/DistributionProperty.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptExpression.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 47 / 52 (90.38%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/PlannerProfile.java|3|4|75.00%|[323]|
|:large_blue_circle:|com/starrocks/sql/optimizer/validate/MVRewriteValidator.java|13|16|81.25%|[39, 81, 82]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|20|21|95.24%|[382]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|4|4|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java [hunk 1]
```diff
@@ -153,6 +153,14 @@ public CostEstimate visitPhysicalOlapScan(PhysicalOlapScanOperator node, Express
             Statistics statistics = context.getStatistics();
             Preconditions.checkNotNull(statistics);
             if (node.getTable().isMaterializedView()) {
+                // If materialized view force rewrite is enabled, hack the materialized view
+                // as zero so can be chosen by the optimizer.
+                ConnectContext ctx = ConnectContext.get();
+                SessionVariable sessionVariable = ctx.getSessionVariable();
+                if (sessionVariable.isEnableMaterializedViewForceRewrite()) {
+                    return CostEstimate.zero();
+                }
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:162]: This is a simple implement to ensure the better chance for materialized view and don't break the optimizer for now.

> how can you deal with enforcer cost if mv rewritten chosen?
> and how to choose between two mvs?

Just decide the choose of the materialized view just by the the plan tree and don't disturb the optimizer for now.
If the two mvs have the same plan tree,  it cannot choose the better one.
If the a mv has a simpler plan tree, the total cost should be smaller and choose it.


Later maybe we can introduce `CostContext` to compare the `GroupExpression` with the `hasRewrittenByMV` flag.

or is there better advise?

**Review by Seaven — APPROVED** [review]

**Review by ABingHuang — APPROVED** [review]

**Pull request merged**